src/main/java/org/truffleruby/core/hash/HashNodes.java

```diff
@@ -886,28 +897,44 @@ protected RubyHash rehashBuckets(RubyHash hash,
             assert HashOperations.verifyStore(getContext(), hash);
 
             final boolean compareByIdentity = byIdentityProfile.profile(hash.compareByIdentity);
+
             final Entry[] entries = (Entry[]) hash.store;
             Arrays.fill(entries, null);
 
             Entry entry = hash.firstInSequence;
-
             while (entry != null) {
                 final int newHash = hashNode.execute(entry.getKey(), compareByIdentity);
                 entry.setHashed(newHash);
                 entry.setNextInLookup(null);
+
                 final int index = BucketsStrategy.getBucketIndex(newHash, entries.length);
                 Entry bucketEntry = entries[index];
 
                 if (bucketEntry == null) {
                     entries[index] = entry;
                 } else {
-                    while (bucketEntry.getNextInLookup() != null) {
+                    Entry previousEntry = entry;
+
+                    do {
+                        if (compareHashKeysNode.equalKeys(
+                                compareByIdentity,
+                                entry.getKey(),
+                                newHash,
+                                bucketEntry.getKey(),
+                                bucketEntry.getHashed())) {
+                            BucketsStrategy.removeFromSequenceChain(hash, entry);
```

Comment:
It seems weird and error-prone that `removeFromSequenceChain()` only does half of it.
I think it should do both, and `DeleteLastNode` should instead use a variant of it (`removeLastFromSequenceChain`, which asserts `entry.getNextInSequence() == null`).